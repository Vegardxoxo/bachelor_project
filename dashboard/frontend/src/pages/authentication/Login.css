#LogIn {
    margin-top: 20%;
    display: flex;
    flex-direction: column;
    align-items: center;
}

#Headline {
    font-size: xx-large;
}