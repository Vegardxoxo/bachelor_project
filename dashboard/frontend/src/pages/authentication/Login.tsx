import {useEffect, useState} from 'react';
import {Box, Button, Container, TextField} from '@mui/material';
import Typography from '@mui/joy/Typography';
import './Login.css'
import {useSetRecoilState} from "recoil";
import {isAuthAtom} from "../../globalVariables/variables";
import {useNavigate} from "react-router-dom";

/**
 * The login page.
 */
export default function Login() {
    const [email, setEmail] = useState('');
    const [password, setPassword] = useState('');
    const [errorMessage, setErrorMessage] = useState('');
    const setIsAuthenticated = useSetRecoilState(isAuthAtom);
    const navigate = useNavigate();

    async function onSubmit(e: { preventDefault: () => void; }) {
        e.preventDefault();
        try {
            const response = await fetch('http://127.0.0.1:8000/api/login/', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    primary_user_email: email,
                    password: password,
                }),
            });

            const data = await response.json();

            if (response.ok) {
                handleAuth(data);
                setIsAuthenticated(true);
                navigate('/');
            } else {
                setErrorMessage(data.detail);
            }
        } catch (error) {
            console.error(error);
        }
    }


    //Stores the access and refresh token in local storage.
    const handleAuth = (data: { access: string, refresh: string }) => {
        localStorage.setItem('access', data.access)
        localStorage.setItem('refresh', data.refresh)

    }

    useEffect(() => {
        const token = localStorage.getItem('access')
        const isLoggedIn = async () => {
            if (token) {
                const response = await fetch('http://127.0.0.1:8000/api/login/verify/', {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                    },
                    body: JSON.stringify({
                        token: token,
                    }),

                })
                if (response.ok) {
                    navigate('/');
                    setIsAuthenticated(true)
                }

            }
        }
        isLoggedIn()

    }, [])


    return (
        <>
            <Container component='main' maxWidth='sm'>
                <Box id="LogIn">
                    <Typography id="Headline">
                        Logg inn
                    </Typography>
                    {errorMessage && <Typography sx={{color: "red"}}>{errorMessage}</Typography>}
                    <Box component='form'
                         onSubmit={onSubmit}
                         sx={{mt: 1}}
                    >
                        <TextField onChange={(e) => {
                            setEmail(e.target.value);
                            setErrorMessage('');
                        }} margin='normal' required fullWidth
                                   id='loginID' label='E-post' data-testid='emaill' name='LoginID' autoFocus
                                   type={"email"}/>
                        <TextField onChange={(e) => {
                            setPassword(e.target.value);
                            setErrorMessage('')
                        }} margin='normal'
                                   required
                                   fullWidth
                                   id='password'
                                   data-testid='passwordd'
                                   label='Passord'
                                   name='password'
                                   type='password'/>
                        <Button type='submit' fullWidth variant='contained' data-testid='buttonn'
                                sx={{mt: 3, mb: 2, backgroundColor: '#005aa7'}}>Logg inn</Button>
                    </Box>
                </Box>
            </Container>
        </>
    );
}
