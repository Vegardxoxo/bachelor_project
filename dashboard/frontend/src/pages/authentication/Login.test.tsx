import React from 'react';
import {cleanup, render, screen, waitFor} from '@testing-library/react';
import '@testing-library/jest-dom';
import Login from './Login';
import 'isomorphic-fetch';
import {RecoilRoot} from 'recoil';
import {MemoryRouter} from 'react-router-dom';
import userEvent from "@testing-library/user-event";
import {act} from "react-dom/test-utils";
import renderer from "react-test-renderer";

describe('Login component', () => {
    beforeEach(async () => {
        jest.spyOn(global, 'fetch').mockImplementation(() =>
            Promise.resolve(
                new Response(JSON.stringify({access: 'fakeAccessToken', refresh: 'fakeRefreshToken'}))
            )
        );

        jest.spyOn(global.localStorage, 'setItem');
        await act(async () => {
            render(
                <RecoilRoot>
                    <MemoryRouter>
                        <Login/>
                    </MemoryRouter>
                </RecoilRoot>
            );
        });
    });

    afterEach(() => {
        cleanup();
        jest.clearAllMocks();
    });

    it('renders login form', () => {
        expect(screen.getByTestId('emaill')).toBeInTheDocument();
    });

    it('submits login form successfully', async () => {
        const emailInput = screen.getByTestId('emaill');
        const passwordInput = screen.getByTestId('passwordd');
        const submitButton = screen.getByTestId('buttonn');
        await act(async () => {
            userEvent.type(emailInput, 'test@example.com');
            userEvent.type(passwordInput, 'password');
            userEvent.click(submitButton);
        });

        await waitFor(() => {
            expect(global.fetch).toHaveBeenCalledTimes(1);
        });
    });

    it('matches snapshot', async () => {
        await act(async () => {
            const testRenderer = renderer.create(
                <RecoilRoot>
                    <MemoryRouter>
                        <Login/>
                    </MemoryRouter>
                </RecoilRoot>
            );
            expect(testRenderer.toJSON()).toMatchSnapshot();
        });
    });
});
