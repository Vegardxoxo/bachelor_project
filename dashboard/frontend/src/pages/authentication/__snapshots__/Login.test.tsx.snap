// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component matches snapshot 1`] = `null`;
